packit/upstream.py

```diff
@@ -921,3 +922,98 @@ def get_version_from_tag(self, tag: str) -> str:
+        if archive_root_dir is None:
+            archive_root_dir = self.get_archive_root_dir_from_template()
+
+        if archive_root_dir is None:
+            raise PackitException("Unable to get archive root dir.")
+
+        return archive_root_dir
+
+    def get_tar_archive_dir(self, archive: str) -> Union[str, None]:
+        """
+        Returns tar archive top-level directory, if there is exactly one.
+
+        :param archive: name of tar/compressed tar archive
+        :return: archive top level directory if exactly one is found
+                None in other cases
+        """
+
+        tar = tarfile.open(f"{self.absolute_specfile_dir}/{archive}")
+        root_dirs = set()
+        for tar_item in tar.getmembers():
+            if tar_item.isdir() and "/" not in tar_item.name:
+                root_dirs.add(tar_item.name)
+            # required for archives where top-level dir was added using tar --transform
+            # option - in that case, tar archive will not contain dir related entry
+            if tar_item.isfile() and "/" in tar_item.name:
+                root_dirs.add(tar_item.name.split("/")[0])
+
+        root_dirs_count = len(root_dirs)
+        archive_root_items_count = len(
+            {i.name for i in tar.getmembers() if "/" not in i.name}
+        )
+
+        if root_dirs_count == 1:
+            root_dir = root_dirs.pop()
+            logger.debug(f"Directory {dir} found in archive {archive}")
+            return root_dir
+        elif root_dirs_count == 0:
+            logger.warning(f"No directory found in archive {archive}.")
+        elif root_dirs_count >= 2:
+            logger.warning(
+                f"Archive {archive} contains multiple directories on the top level: "
+                f"the common practice in the industry is to have only one in the "
+                f'following format: "PACKAGE-VERSION"'
+            )
+        elif archive_root_items_count > 1:
+            logger.warning(
+                f"Archive f{archive} contains multiple root items. It can be "
+                f"intentional or can signal incorrect archive structure."
+            )
+        return None
+
+    def get_archive_root_dir_from_template(self) -> Union[str, None]:
+        """
+        Generates archive root dir based on archive_root_dir_template.
+        archive_root_dir_template default value is "{upstream_pkg_name}-{version}"
+
+        :returns: archive root dir name based on template
+        """
+        template = self.package_config.archive_root_dir_template
+        if not isinstance(template, str) or not template:
+            logger.error(
+                "Incorrectly configured archive_root_dir. It have to be non-empty string."
+            )
+            return None
+
+        archive_root_dir = self.package_config.archive_root_dir_template.replace(
+            "{upstream_pkg_name}", self.package_config.upstream_package_name
+        ).replace("{version}", self.get_version())
```

Comment:
please use the formatting syntax instead: `self.package_config.archive_root_dir_template.format(upstream_pkg_name=..., version=...)`

Reply:
but `.format` will fail if no value is provided for any of `{}`

```python
>>> " {a}, {b}".format(a=4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'b'
>>> 
```
I know the probability is really low, but if someone will have `{.*}` in the directory name, `.format()`, will fail, so I will prefer to keep `.replace()`

Reply:
ah, that makes sense, haven't realized that